### PR TITLE
Robyn Marowitz LiDAR pull request

### PR DESCRIPTION
**Pull request by rmarowitz**
# Pull Request Template

## CI Checks
- [ ] The notebook runs from start to finish on all operating systems:
   - [ ] Mac
   - [ ] Windows
   - [ ] Linux 
### Reproducibility
- [ ] Are the data downloaded in the code
- [ ] Are paths created to ensure they work on all operating systems (using  os.path.join)
- [ ] Are comments used to clarify the contents of the code that can’t be clarified using expressive variable and function names alone? (not too many comments - just enough)
- [ ] Does the notebook run from start to finish?

### Functions
- [ ] Do functions follow PEP 8 format conventions?
- [ ] Are function docstrings clear (all  inputs and outputs clearly  described and defined)
- [ ] Are function names expressive (the name describes what the function does)?
- [ ] Are functions easy to understand and read?
- [ ] How many tasks does each function do? (ideally a function does one thing well).
### Package imports
- [ ] Are standard modules (those included with the base  python install) vs. third party (related but externally developed tools) import groups correct with appropriate spacing in between each group?
- [ ] Are variable names throughout the code, expressive?
- [ ] Suggest changes if not, highlight what was done well
- [ ] Is the code overall easy to understand and read? Are there things that would make it more clear / cleaner?

### DRY Code
- [ ] Are segments of code repeated in the file or is the code DRY?
- [ ] Are loops used to optimize DRY code?
- [ ] Are functions used to optimize DRY code?
- [ ] Are there any areas that could be potentially improved (you can suggest improvements OR you can just highlight parts of the code where you suspect it could  be cleaner / more efficient.

### Novel Approaches to Problem solving
- [ ] Highlight any novel approaches to completing the assignment.

### CODE REVIEW

**Comment by eculler** [comment]
@ryanboarman - Please review this PR #50 (Canvas isn't working so let's try GitHub directly!)

**Comment by rmarowitz** [comment]
Reviewer, 

I know there are tons of commits, I looked at this a few weeks ago thinking it was due and initially added NDVI files. Now it should be as it's meant to.

@eculler I also noticed that I still have those CI checks on here since I created this before you made that change Wednesday and I was unable to create a pull request to merge in your 2 updates.

Thanks, 
Robyn

File: clean.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import pandas as pd
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import rasterstats as rs
+import xarray as xr
+import seaborn as sns
+
+class NEONDataLoader(object):
+    """Parent class to load NEON height data."""
+
```
Comment (ryanboarman) [inline clean.py:14]: Docstring is missing parameters and returns

File: clean.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        sns.regplot(data=self.height_stats, x='lidar_mean',y='insitu_mean', 
+                   color='purple',
+                   ax=ax2)
+        ax2.set(xlabel='Mean Lidar Height(m)',ylabel='Mean Insitu Height(m)',
+               title=('Plot 2: Lidar Mean vs Insitu Mean Tree Height\n'
+                     '{} Field Site').format(self.site_name),
+                ylim=(0,30),
+                xlim=(0,30))
+        ax2.plot((0,1), (0,1), transform=ax2.transAxes, ls='--', c='k')
```
Comment (ryanboarman) [inline clean.py:132]: add more # comments to clarify the code

File: clean.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        '{base_dir}', '2013',
+        'lidar',
+        '{site_name_low}_lidarCHM.tif')
+    
+    site_name = NotImplemented
+    id_col_name = NotImplemented
+    formatting_dict = NotImplemented
+    id_modifier = None
+    
+    def __init__(self):
```
Comment (ryanboarman) [inline clean.py:37]: Docstring is missing parameters and returns

File: clean.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        
+        self._insitu_height_stats = None
+        self._lidar_chm_stats = None
+        self._height_stats = None
+        
+
+    @property
+    def lidar_chm_stats(self):
+        """
+        Calculate max and mean tree height from LiDAR.
```
Comment (ryanboarman) [inline clean.py:55]: Docstring is missing parameters and returns

File: clean.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                inplace=True)
+            if not self.id_modifier is None:
+                self._lidar_chm_stats[self.id_col_name] = (
+                    self._lidar_chm_stats[self.id_col_name].apply(self.id_modifier))
+        return self._lidar_chm_stats
+    
+    @property
+    def insitu_height_stats(self):
+        """
+        Calculate insitu tree height data max and mean.
```
Comment (ryanboarman) [inline clean.py:82]: Docstring is missing parameters and returns

File: clean.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                          .groupby('plotid')
+                          .stemheight
+                          .agg(['max','mean'])
+                         .rename(columns={
+                                 'max':'insitu_max', 'mean':'insitu_mean'}))
+        return self._insitu_height_stats 
+    
+    @property
+    def height_stats(self):
+        """
```
Comment (ryanboarman) [inline clean.py:95]: Docstring is missing parameters and returns

File: clean.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                self.lidar_chm_stats
+                .merge(self.insitu_height_stats,
+                       right_index=True,
+                       left_on=self.id_col_name))
+
+        return self._height_stats
+    
+    def plots(self):
+        """
+        Create plots with regression lines.
```
Comment (ryanboarman) [inline clean.py:109]: Docstring is missing parameters and returns

File: marowitz_robyn_lidar_uncertainty.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import earthpy as et 
+import pandas as pd
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import rasterstats as rs
+import xarray as xr
+import rioxarray as rxr
+import seaborn as sns
+
+import clean
```
Comment (ryanboarman) [inline marowitz_robyn_lidar_uncertainty.py:33]: good job on the imports.  Local application/library specific imports should be last. You may want to put them in alphabetical order though

File: marowitz_robyn_lidar_uncertainty.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+class SOAPDataLoader(clean.NEONDataLoader):
+    site_name = 'SOAP'
+    id_col_name = 'ID'
+    formatting_dict = {'seperator':'-', 'plot': ''}
+    
+    def id_modifier(self, id):
+        return 'SOAP' + str(id)
+
+
+soap_data_loader = SOAPDataLoader()
```
Comment (ryanboarman) [inline marowitz_robyn_lidar_uncertainty.py:60]: good expressive names

File: marowitz_robyn_lidar_uncertainty.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+# In[10]:
+
+
+soap_data_loader.plots()
+
+
+# In[ ]:
+
+
+
```
Comment (ryanboarman) [inline marowitz_robyn_lidar_uncertainty.py:89]: overall the code is DRY

File: clean.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        """
+    
+        fig, (ax1,ax2) = plt.subplots(1, 2, figsize=(14, 7))
+        # ax1.scatter(sjer_df.lidar_mean, sjer_df.insitu_mean)
+        sns.regplot(data=self.height_stats, x='lidar_max',y='insitu_max', 
+                   color='purple',
+                   ax=ax1)
+        ax1.plot((0,1), (0,1), transform=ax1.transAxes, ls='--', c='k')
+        ax1.set(xlabel='Max Lidar Height(m)',ylabel='Max Insitu Height(m)',
```
Comment (ryanboarman) [inline clean.py:118]: @rmarowitz , add more detail on what the function does by including # comments and more information in the docstrings

File: clean.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+import earthpy as et
+import pandas as pd
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import rasterstats as rs
+import xarray as xr
+import seaborn as sns
+
+class NEONDataLoader(object):
```
Comment (ryanboarman) [inline clean.py:12]: Nice job creating a class that works!!

**Comment by ryanboarman** [comment]
- [x ]  The notebook runs from start to finish on all operating systems:
- [x ]  Mac
- [ ]  Windows
- [ ]  Linux
-
 Reproducibility

- [x ]  Are the data downloaded in the code
- [x ]  Are paths created to ensure they work on all operating systems (using os.path.join)
- [ ]  Are comments used to clarify the contents of the code that can’t be clarified using expressive variable and function names alone? (not too many comments - just enough)
- [x]  Does the notebook run from start to finish?

PEP 8 standards & Code Readability
- Functions
- [x ]  Do functions follow PEP 8 format conventions?
- [ ]  Are function docstrings clear (all inputs and outputs clearly described and defined)
- [x ]  Are function names expressive (the name describes what the function does)?
- [x ]  Are functions easy to understand and read?
- [x ]  How many tasks does each function do? (ideally a function does one thing well).
- [x ] Package imports

- [ x]  Are standard modules (those included with the base python install) vs. third party (related but externally developed tools) import groups correct with appropriate spacing in between each group?
- [x ]  Are variable names throughout the code, expressive?
- [ x]  Suggest changes if not, highlight what was done well
- [x ]  Is the code overall easy to understand and read? Are there things that would make it more clear / cleaner?
- DRY Code

- [x ]  Are segments of code repeated in the file or is the code DRY?
- [x ]  Are loops used to optimize DRY code?
- [ x]  Are functions used to optimize DRY code?
- [x ]  Are there any areas that could be potentially improved (you can suggest improvements OR you can just highlight parts of the code where you suspect it could be cleaner / more efficient.
- [ x] Novel Approaches to Problem solving

- [x ]  Highlight any novel approaches to completing the assignment.

**Review by ryanboarman — COMMENTED** [review]
In general good job. Nice that you made a class and it worked out. Only main suggestions I would say is to add more information on the docstrings and add more #comments throughout the document to make it easier to read. @rmarowitz

**Comment by rmarowitz** [comment]
@ryanboarman Thanks for the detailed review :)

File: clean.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
+import os
+import pathlib
+
+import earthpy as et
+import pandas as pd
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import rasterstats as rs
+import xarray as xr
```
Comment (StepanBryleev) [inline clean.py:9]: It seems that xarray library is not needed for this task and not used.

File: clean.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+            # Calculate the zonal stats
+            chm_stats = rs.zonal_stats(
+                plots_gdf, 
+                self.chm_path,
+                stats=['mean', 'max'],
+                geojson_out=True,
+                nodata=0,
+                copy_properties=True
+            )
+            self._lidar_chm_stats = gpd.GeoDataFrame.from_features(chm_stats)
```
Comment (StepanBryleev) [inline clean.py:70]: You could add more # comments to describe data processing

File: marowitz_robyn_lidar_uncertainty.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+import os
+import pathlib
+
+import earthpy as et 
+import pandas as pd
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import rasterstats as rs
+import xarray as xr
```
Comment (StepanBryleev) [inline marowitz_robyn_lidar_uncertainty.py:29]: It seems that xarray and rioxarray libraries are not needed for this task and not used.

File: marowitz_robyn_lidar_uncertainty.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+
+soap_data_loader.plots()
+
+
+# In[ ]:
+
+
+
+
```
Comment (StepanBryleev) [inline marowitz_robyn_lidar_uncertainty.py:90]: Good job!

File: marowitz_robyn_lidar_uncertainty.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+# In[9]:
+
+
+sjer_data_loader.plots()
+
+
+# In[10]:
+
+
+soap_data_loader.plots()
```
Comment (StepanBryleev) [inline marowitz_robyn_lidar_uncertainty.py:83]: You could add more # comments to describe your final workflow.

**Review by StepanBryleev — COMMENTED** [review]
# Pull Request Template

## Review Checklist

### Reproducibility
- [x] Are the data downloaded in the code
- [x] Are paths created to ensure they work on all operating systems (using  os.path.join)
- [ ] Are comments used to clarify the contents of the code that can’t be clarified using expressive variable and function names alone? (not too many comments - just enough)
- [x] Does the notebook run from start to finish?

### PEP 8 standards & Code Readability

### Functions
- [x] Do functions follow PEP 8 format conventions?
- [ ] Are function docstrings clear (all  inputs and outputs clearly  described and defined)
- [x] Are function names expressive (the name describes what the function does)?
- [x] Are functions easy to understand and read?
- [x] How many tasks does each function do? (ideally a function does one thing well).

### Package imports
- [ ] Are standard modules (those included with the base  python install) vs. third party (related but externally developed tools) import groups correct with appropriate spacing in between each group?
- [x] Are variable names throughout the code, expressive?
- [x] Is the code overall easy to understand and read? Are there things that would make it more clear / cleaner?

### DRY Code
- [x] Is the code DRY?
- [ ] Are loops used to optimize DRY code?
- [x] Are functions used to optimize DRY code?
- [x] Are there any areas that could be potentially improved (you can suggest improvements OR you can just highlight parts of the code where you suspect it could  be cleaner / more efficient.

### Novel Approaches to Problem solving
- [x] Highlight any novel approaches to completing the assignment.

You created a class and it works perfect! Good job @rmarowitz !

**Comment by eculler** [comment]
Great work, Robyn! Excellent use of classes and a module. Your docstrings could use some work - you're missing the parameters, attributes and returns. Make sure to document your work.

**Pull request closed without merging**